### PR TITLE
[red-knot] follow simple assignments

### PR DESCRIPTION
**Pull request by carljm**
## Summary

Follow simple assignments in type resolution.

## Test Plan

cargo test

### CODE REVIEW

File: crates/red_knot/src/symbols.rs
Comment (MichaReiser) [inline crates/red_knot/src/symbols.rs:539]: Take is mainly used when you want to do something with the return value

```suggestion
                self.current_definition = None
```

File: crates/red_knot/src/symbols.rs
Comment (MichaReiser) [inline crates/red_knot/src/symbols.rs:447]: How expensive to we expect `Definition` becomes. Let's say we have a function definition with all its parameters (all have a name). The worst case is that this allocates a lot of new strings. Shoul definition be wrapped in an `Rc`? Or should `definitions` be interned in an `Arena` so that we can reference definitions by ID instead?

What are the cases where we have multiple definitions out of a single enclosing definition? Should we use `take` here?

File: crates/red_knot/src/symbols.rs
Comment (MichaReiser) [inline crates/red_knot/src/symbols.rs:537]:
```suggestion
                self.current_definition = Some(Definition::Assignment(TypedNodeKey::from_node(node))));
```

File: crates/red_knot/src/types/infer.rs
Comment (MichaReiser) [inline crates/red_knot/src/types/infer.rs:88]: I can see how resolving definitions might become expensive :laughing:

**Review by MichaReiser — APPROVED** [review]

File: crates/red_knot/src/symbols.rs
Comment (carljm) [reply crates/red_knot/src/symbols.rs:447]: My intention is for Definition to stay quite small, because in most cases it should just hold a `TypedNodeKey` -- we don't want to copy lots of stuff from the AST when we can just reference the AST instead.

The case for shared Definition is something like `x = y = z`, or even more complex (why we need the nested visit), `(x, (y, z)) = foo`.

Ideally I might like to only reference the `Expr` on the RHS here (or even just part of it, for the unpacking case), but we have the general problem that we can't have a `TypedNodeKey` to arbitrary `Expr`.

File: crates/red_knot/src/types/infer.rs
Comment (carljm) [reply crates/red_knot/src/types/infer.rs:88]: Yes, it would be awesome if we could directly reference AST nodes instead of having to search to resolve :/

File: crates/red_knot/src/symbols.rs
Comment (MichaReiser) [reply crates/red_knot/src/symbols.rs:447]: We can have a `TypedNodeKey` to arbitrary expressions. You can use `ExpressionRef`

**Comment by MichaReiser** [comment]
You know have to label your PRs or they'll show up in the changelog :laughing:

File: crates/red_knot/src/symbols.rs
Comment (carljm) [reply crates/red_knot/src/symbols.rs:447]: Ooh very good, I'll need to look at that in more detail.

**Comment by carljm** [comment]
Ugh, yes I know, I did remember to label one of them today! https://github.com/astral-sh/ruff/pull/11176

**Comment by carljm** [comment]
I think I would have remembered before merging :)

File: crates/red_knot/src/symbols.rs
Comment (carljm) [reply crates/red_knot/src/symbols.rs:447]: It looks like using `ExpressionRef` _should_ be possible (it can be cast to `AnyNodeRef`), but it will take a bit more investigation, because `ExpressionRef` doesn't impl `AstNode`, and currently `TypedNodeKey` only works for `N: AstNode`. Not going to hold up this PR on figuring this out, but I'll add a TODO to look at it later.

Comment (carljm) [reply crates/red_knot/src/symbols.rs:447]: Actually on second thought as I try to write the TODO, I don't even think we can / want to store just Exprs for Assignments. The RHS of an unpacking assigment may be an expression that returns a sequence type, but isn't a tuple or list literal, so we can't necessarily pick it apart to its component bits at AST level. We will need the whole LHS and the whole RHS to do the unpacking at type level, and so effectively we need the whole statement. So although there may be other cases where we'd like to reference an Expr, this isn't actually one of them.

**Pull request merged**
